default/AI/FleetUtilsAI.py

```diff
@@ -37,6 +37,23 @@ def count_parts_fleetwide(fleet_id, parts_list):
     return tally
 
 
+def count_troops_in_fleet(fleet_id):
+    """
+    :param fleet_id:
+    :return: total troopCapacity of the fleet
+    """
+    universe = fo.getUniverse()
+    fleet = universe.getFleet(fleet_id)
+    if not fleet:
+        return 0
+    fleet_troop_capacity = 0
+    for ship_id in fleet.shipIDs:
+        ship = universe.getShip(ship_id)
+        if ship:
```

Comment:
Hmm, a good question, really a policy decision and one I think merits a bit of discussion.  I think that when I first started working on this our Universe code was much more buggy/unreliable and the AI would sometimes wind up with invalid object ids (at least, ids for objects that it would not be able to get even a bare-boned copy of); trying to retrieve an object with such would return a null pointer (and now not quite exactly a null pointer but still something largely equivalent).  So at that time the Universe issues would make themselves known plenty through the human client and we didn't really want the AI crashing from such things.

At this point, the AI should (*at least for the most part, see below) not be winding up with object IDs that it cannot retrieve an object for.  Certainly in this case, the ids in fleet.shipIDs should all return valid objects that the AI client can retrieve.  If one was invalid, that indicates a bug that should be brought to the attention of the user.  I suspect that a hard AI crash is probably not what we want in that case, though-- we'd want to catch the exception and flag the problem to the user with the debugging tools you've put together.   I think we have those wrappers in place at all the high level functions, but we may want to also wrap some of the lower level functions so that a smaller portion of AI operation would be impacted.

One important side note is that depending on context, the AI may need to double check that the object id is not in its list of 'known destroyed objects'.  Here, depending on where the code calling count_troops_in_fleet() had gotten the fleet_id from, it should perhaps check that for the fleet_id (for example, if it was from some list saved from a previous turn).  As for the fleet.shipIDs, I am pretty sure that at this point none of them should be known to be destroyed or they wouldn't be in that list/vector (but I have not dug through the code to confirm that).

The exception I flag above about valid ids is that once the AI starts processing sitreps, it may from them get object ids for which it has no visibility whatsoever, and for which it would get a null object if it tried retrieving it.  That would not be a bug and we should not crash from it.  So we should keep in mind where we are getting the object id from, and in some cases may need to still be careful about not assuming the id is valid.  At the moment though, no other source for potentially 'invalid' ids comes to mind other than sitreps.